### PR TITLE
Feature/api specification

### PR DESCRIPTION
**Pull request by pbolduc**
Added OpenAPI 3.0 for designing API.   I based this on the default Petstore created by Swagger Editor, so there can be some _pet_ references still.

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  description: "Access to projects"
+- name: "user"
+  description: "Operations about user"
+  externalDocs:
+    description: "Find out more about our project"
+    url: "https://github.com/bcgov/maltd"
+schemes:
+- "https"
+paths:
+  /users:
```
Comment (akroon3r) [inline swagger.yaml]: Do we need this endpoint? I thought it was 1-1 exact match?

File: swagger.yaml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      code:
+        type: "integer"
+        format: "int32"
+      type:
+        type: "string"
+      message:
+        type: "string"
+externalDocs:
+  description: "Find out more about Swagger"
+  url: "http://swagger.io"
```
Comment (akroon3r) [inline swagger.yaml]: missing newline at end of file

File: swagger.yaml (hunk 1 above)
Comment (akroon3r) [reply swagger.yaml]: ie// is there a case where we need to list ALL users?

Comment (pbolduc) [reply swagger.yaml]: We would not list all users. This endpoint I had added for the search, 

```
/users?q=akroon3r
```

which you could get the details about a user with

```
/users/akroon3r
```

File: swagger.yaml (hunk 2 above)
Comment (pbolduc) [reply swagger.yaml]: Think that was from the OAS 2.0 document, the new doc ends with the schema section

File: swagger.yaml (hunk 1 above)
Comment (akroon3r) [reply swagger.yaml]: thanks for clarifying :)

File: swagger.yaml (hunk 2 above)
Comment (akroon3r) [reply swagger.yaml]: 100% correct, my bad.

**Comment by pbolduc** [comment]
> will we have/should we include any header information in the requests which would probably be used for logging purposes? such as IDIR, etc?

How is IDIR authentication handled? Do we get a JWT token?  Typically in JWT, there at attributes about the current user's login name, etc.

**Comment by sdevalapurkar** [comment]
> > will we have/should we include any header information in the requests which would probably be used for logging purposes? such as IDIR, etc?
> 
> How is IDIR authentication handled? Do we get a JWT token? Typically in JWT, there at attributes about the current user's login name, etc.

Yeah I am not too sure at the moment, but it was just something I wanted to check and keep in mind, not a blocker for this PR

**Comment by akroon3r** [comment]
@ChrisHoban please shine your knowledge flood light upon this PR

**Comment by pbolduc** [comment]
Note, I was using [Swagger Editor](https://swagger.io/tools/swagger-editor/) to edit and view the API.  I assume we will be able to use [SwaggerCodegen](https://swagger.io/tools/swagger-codegen/) to generate javascript library (Fetch)?

image

**Comment by ChrisHoban** [comment]
https://www.keycloak.org/docs-api/5.0/rest-api/index.html#_users_resource :)

File: swagger.yaml (hunk 1 above)
Comment (pbolduc) [reply swagger.yaml]: We could combine the user get and search.  There are pros and cons.  It depends on if the user get also returns all the projects they are assigned to. In this case could be slower.  Having a separate search could allow, in the future, a automatic background load of the the project details. The look up of the user should be quick.

**Pull request merged**